### PR TITLE
Fixes some NanoUI machines not updating on button press

### PR DESCRIPTION
**Pull request by 9600bauds**
FUCK
NANO

### CODE REVIEW

**Comment by PJB3005** [comment]
What?

Are you sure that returning 1 in `Topic()` does anything at all?

**Comment by 9600bauds** [comment]
Yes.
You need to return 1 in NanoUI topic() otherwise it won't update on buttonpress.
I never knew this before.

**Comment by PJB3005** [comment]
But that's impossible?

How could nano possibly intercept the return 1 in this case?

**Comment by 9600bauds** [comment]
it

just

werks

**Comment by PJB3005** [comment]
Have you tested it?

**Comment by 9600bauds** [comment]
Of course, like I said, it just werks

**Comment by clusterfack** [comment]
How did you find this out? I'm also skeptical

**Comment by optimumtact** [comment]
nanoui's call the topic of the machine from their own topic

https://github.com/d3athrow/vgstation13/blob/Bleeding-Edge/code/modules/nano/nanoui.dm#L498

basically all links are proxied through the nano datum for that ui, see line 520 for the exact call

edit:
something you may want to consider is in our nanoui update neersighted changed it to call a non topic proc ui_act, explicitly indicating the interception that nano is doing (it's non obvious otherwise)

**Comment by 9600bauds** [comment]
I found this out because canisters and portascrubbers had almost exactly the same Topic() and UI code, yet one of them auto-updated on click and the other one didn't.
I couldn't make any sense of it, so I decided to try random shit.
The first random thing I tried was adding a 1 to the return of portascrubbers, since the canister returned 1 rather than nothing.
And it worked.
And I found out the reason why Xenobotany machines used to not update on buttonpress: Their topic() didn't return 1. I had fixed this earlier by just making it update in every single topic() call, which is the improper solution.
I then tried this one last time in the Chemmaster, which also does not update on buttonpress: It didn't return 1. I made it return 1. Now it werks.

**Comment by PJB3005** [comment]
TL;DR: NanoUI is snowflake bullshit?

**Comment by optimumtact** [comment]
Snowflake?

it redirects the topic so it can handle a lot of the status/distance checks in once place and apply them consistently

nothing wrong with that as a design goal.

**Comment by PJB3005** [comment]
Returning 1 in `Topic()` should _not_ be what causes the update though.

**Comment by PsiOmegaDelta** [comment]
In `nanoui.dm`, Bay flavor mind you, see the last lines of code:

``` DM
/datum/nanoui/Topic(href, href_list)
    update_status(0) // update the status
    if (status != STATUS_INTERACTIVE || user != usr) // If UI is not interactive or usr calling Topic is not the UI user
        return

    // This is used to toggle the nano map ui
    var/map_update = 0
    if(href_list["showMap"])
        set_show_map(text2num(href_list["showMap"]))
        map_update = 1

    if(href_list["mapZLevel"])
        set_map_z_level(text2num(href_list["mapZLevel"]))
        map_update = 1

    // This is the line where the return value of the Topic call matters.
    // Topic is assumed to return 1 if the call was handled to keep the number of updates down.
    if ((src_object && src_object.Topic(href, href_list, state)) || map_update)
        nanomanager.update_uis(src_object) // update all UIs attached to src_object
```

This is why returning 1 makes a difference. Whether it should is another matter.

**Comment by clusterfack** [comment]
Alright that makes sense then, please make sure none of these nanouis have autoupdating enabled if they don't need to anymore then bauds

**Comment by 9600bauds** [comment]
I can't find anything that would make them autoupdating

**Comment by clusterfack** [comment]
The ui.set_auto_update(1) in the initial ui_interact proc sets them to auto update

**Comment by 9600bauds** [comment]
Will fix tomorrow

**Comment by 9600bauds** [comment]
This seems to make canisters update a little better, too
